Rework the tutorial (#108)
scrapinghub/web-poet: docs/intro/tutorial.rst

Within ``tutorial-project/tutorial/pages``, create:

-   An ``__init__.py`` file.

-   A ``books_toscrape_com.py`` file, for page object class code targeting
    `books.toscrape.com`_.

Your project folder should look as follows:

.. code-block:: text

    tutorial-project/
        run.py
        tutorial/
            __init__.py
            items.py
            pages/
                __init__.py
                books_toscrape_com.py

Create an item class
====================

While it is possible to store the extracted data in a Python dictionary, it is
a good practice to create an item class that:

-   Defines the specific attributes that you aim to extract, triggering an
    exception if you extract unintended attributes or fail to extract expected
    attributes.

-   Allows defining default values for some attributes.

web-poet uses itemadapter_ for item class support, which means that any kind of
item class can be used. In this tutorial, you will use attrs_ to define your
item class.

.. _attrs: https://www.attrs.org/en/stable/
.. _itemadapter: https://github.com/scrapy/itemadapter

Copy the following code into ``tutorial-project/tutorial/items.py``:

> .. literalinclude:: /../tutorial-project/tutorial/items.py

Reviewer: What do you think about moving tutorial-project to the docs folder? The code looks a bit strange as a general-purpose example (imports in the middle of files, etc).

   :language: python
   :lines: 1-6

This code defines a ``Book`` item class, with a single, required, string
``title`` attribute to store the book title.

``Book`` is a minimal class designed specifically for this tutorial. In real

> web-poet projects, you will often define item classes with many more

Reviewer: ```suggestion
web-poet projects, you will usually define item classes with many more
```

    ``BookPage.title`` returns gets mapped to ``Book.title``.

Use your page object class
==========================

Now that you have a page object class defined, it is time to use it.

First, install requests_, which is required by ``web_poet.example``.

.. _requests: https://requests.readthedocs.io/en/latest/user/install/

Then copy the following code into
``tutorial-project/run.py``:

.. literalinclude:: /../tutorial-project/run.py
   :language: python
   :lines: 1-5, 7-8

Execute that code:

..  code-block:: bash

    python tutorial-project/run.py

And the ``print(item)`` statement should output the following:

.. code-block:: python

   Book(title='The Exiled')

In this tutorial you use ``web_poet.example.get_item``, which is a simple,
incomplete implementation of the web-poet standard, built specifically for this
tutorial, for demonstration purposes. In real projects, use instead an actual
web-poet framework, like `scrapy-poet`_.

``web_poet.example.get_item`` serves to illustrate the power of web-poet. Once
you have defined your page object class, a web-poet framework only needs 2
inputs from you: the URL from which you want to extract data, and the import

> path of the Python modules where you define your page object classes.

Reviewer: I think we should require 2 inputs:

* URL
* Item type.

Import path to modules should be something which is handled on setup time, not something which is passed to every get_item call. That's global shared state (modules currently imported in the Python process), and we should be explicit about it; it's not good that a call to get_item may implicitly affect other get_item calls.

As for asking to specifiy the item class, it could be error-prone otherwise. For example, the modules may contain page objects which extract different item types from the same pages. Let me elaborate on this.

I think it's quite common that you can't write a specific URL pattern for an item. So, handle_urls could be annotated just with a domain name, without trying to limit it further by path, etc, and that's perfectly fine. In most cases the crawling code knows what kind of content to expect after following the link, so it should be able to use right page objects on right URLs; it may depend no only on the URL, but also on CSS/XPath of the link extracted. 

While being more precise with handle_urls is fine, it may also backfire sometimes, because this logic can get split between crawling (link extraction code) and handle_urls. 

The main purpose for using URL patterns in handle_urls is to distinguish between different page objects which return the same item class, not to infer the item class returned for an URL.

.. _intro-tutorial:
.. _page-objects:

========
Tutorial
========

> In this tutorial you will learn to use web-poet as you write web data

Reviewer 2: ```suggestion
In this tutorial, you will learn to use web-poet as you write web data
```

Author: I’m not sure about this one.

extraction code for book detail pages from `books.toscrape.com`_.

.. _books.toscrape.com: http://books.toscrape.com/

To follow this tutorial you must first be familiar with Python_ and
:ref:`install web-poet <install>`.

.. _Python: https://docs.python.org/

> Create a project folder

Author: I’m applying the change, but for the record I think it is OK to use the word folder even when there is no GUI involved, and I think it can be less confusing to some users.

==========================

web-poet does not limit how you structure your web-poet web data extraction
code, beyond the limitations of Python itself.

However, in this tutorial you will use a specific project directory structure
designed with web-poet best practices in mind. Consider using the same project

> directory structure in all your web-poet projects.

Reviewer: I think the structure might depend on the framework used, e.g. it might be different in case of scrapy-poet. Should we say "similar" instead of "the same"?

Reviewer: The structure looks fine overall, but I wonder what run.py method should contain in user projects.

Author: > I wonder what run.py method should contain in user projects.

Nothing, it is described as “a file specific to this tutorial” on its first mention.

First create your project folder: ``tutorial-project``.

> Within ``tutorial-project``, create:

Author: Went with including the word folder. I actually like using `/` for this purpose, but I would then use it everywhere, and it felt too verbose to use it in conjugation with the world folder in some of the other references in this section.

-   A ``run.py`` file, a file specific to this tutorial where you will put code
    to test the execution of your web data extraction code.

-   A ``tutorial`` folder, where you will place your web data extraction code.

Within ``tutorial-project/tutorial``, create:

-   An ``__init__.py`` file, so that the ``tutorial`` folder becomes an
    importable Python module.

-   An ``items.py`` file, where you will define item classes to store extracted
    data.

-   A ``pages`` folder, where you will define your page object classes.

Within ``tutorial-project/tutorial/pages``, create:

-   An ``__init__.py`` file.

-   A ``books_toscrape_com.py`` file, for page object class code targeting
    `books.toscrape.com`_.

Your project folder should look as follows:

.. code-block:: text

    tutorial-project/
        run.py
        tutorial/
            __init__.py
            items.py
            pages/
                __init__.py

>                 books_toscrape_com.py

Reviewer 2: What do you think about using GNU tree's formatting on this?

```
.
└── tutorial-project
    ├── run.py
    └── tutorial
        ├── __init__.py
        ├── items.py
        └── pages
            ├── __init__.py
            └── books_toscrape_com.py
```

Create an item class
====================

While it is possible to store the extracted data in a Python dictionary, it is
a good practice to create an item class that:

-   Defines the specific attributes that you aim to extract, triggering an
    exception if you extract unintended attributes or fail to extract expected
    attributes.

-   Allows defining default values for some attributes.

web-poet uses itemadapter_ for item class support, which means that any kind of
item class can be used. In this tutorial, you will use attrs_ to define your
item class.

.. _attrs: https://www.attrs.org/en/stable/
.. _itemadapter: https://github.com/scrapy/itemadapter

Copy the following code into ``tutorial-project/tutorial/items.py``:

.. literalinclude:: /tutorial-project/tutorial/items.py
   :language: python
   :lines: 1-6

This code defines a ``Book`` item class, with a single required ``title``
string attribute to store the book title.

``Book`` is a minimal class designed specifically for this tutorial. In real

> web-poet projects, you will usually define item classes with many more

Reviewer 2: Do you think if we can mention (as an example) and add a link to zyte-common-items here?

website, `books.toscrape.com`_, item classes are usually meant to be usable for
many different websites that provide data with a similar data schema.

Create a page object class
==========================

To write web data extraction code with web-poet, you write :ref:`page object
classes <page-objects>`, Python classes that define how to extract data from a
given type of input, usually some type of webpage from a specific website.

In this tutorial you will write a page object class for webpages of
`books.toscrape.com`_ that show details about a book, such as these:

-   http://books.toscrape.com/catalogue/the-exiled_247/index.html
-   http://books.toscrape.com/catalogue/when-we-collided_955/index.html
-   http://books.toscrape.com/catalogue/set-me-free_988/index.html

Copy the following code into
``tutorial-project/tutorial/pages/books_toscrape_com.py``:

.. literalinclude:: /tutorial-project/tutorial/pages/books_toscrape_com.py
   :language: python
   :lines: 1-11

In the code above:

-   You define a page object class named ``BookPage`` by subclassing
    :class:`~web_poet.pages.WebPage`.

    It is possible to create a page object class subclassing instead the

>     simpler :class:`~web_poet.pages.ItemPage` class, but

Reviewer 2: ```suggestion
    simpler :class:`~web_poet.pages.ItemPage` class. However, the
```

    :class:`~web_poet.pages.WebPage`:

>     -   Indicates that your page object class requires an HTTP response as

Reviewer 2: ```suggestion
    -   Indicates that your page object class requires an :class:`web_poet.page_inputs.http.HttpResponse` as
```

        input, which gets stored in the
        :attr:`~web_poet.pages.WebPage.response` attribute of your page object
        class.

    -   Provides attributes like :attr:`~web_poet.pages.WebPage.html` and
        :attr:`~web_poet.pages.WebPage.url`, and methods like
        :meth:`~web_poet.pages.WebPage.css`,
        :meth:`~web_poet.pages.WebPage.urljoin`, and
        :meth:`~web_poet.pages.WebPage.xpath`, that make it easier to write
        web data extraction code.

-   ``BookPage`` declares ``Book`` as its return type.

    :class:`~web_poet.pages.WebPage`, like its parent class
    :class:`~web_poet.pages.ItemPage`, is a :ref:`generic class <generics>`
    that accepts a type parameter. Unlike most generic classes, however, the
    specified type parameter is used for more than type hinting: it determines

>     the item class that is used to store the data that fields return.

Reviewer 2: Let's also mention that the `.to_item()` method that `WebPage` inherits from `ItemPage` is responsible for returning the said `Book` item.

Author: I am unsure about this. It is covered in the following section, in the 3rd bullet point at the end, and I think that may be the best point to cover this.

> -   ``BookPage`` is decorated with :meth:`~web_poet.overrides.handle_urls`,

Reviewer: the link to handle_urls doesn't work; probably it should point to .rules module

    which indicates for which domain ``BookPage`` is intended to work.

    It is possible to specify more specific URL patterns, instead of only the
    target URL domain. However, the URL domain and the output type (``Book``)
    are usually all the data needed to determine which page object class to
    use, which is the goal of the :meth:`~web_poet.overrides.handle_urls`
    decorator.

-   ``BookPage`` defines a field named ``title``.

    :ref:`Fields <fields>` are methods of page object classes, preferably async
    methods, decorated with :meth:`~web_poet.fields.field`. Fields define the
    logic to extract a specific piece of information from the input of your
    page object class.

    ``BookPage.title`` extracts the title of a book from a book details
    webpage. Specifically, it extracts the text from the first ``h1`` element
    on the input HTTP response.

>     Here ``title`` is not an arbitrary name, it was chosen specifically to

Reviewer 2: ```suggestion
    Here, ``title`` is not an arbitrary name. It was chosen specifically to
```

    match ``Book.title``, so that during web data extraction the value that
    ``BookPage.title`` returns gets mapped to ``Book.title``.

Use your page object class
==========================

Now that you have a page object class defined, it is time to use it.

First, install andi_ and requests_, which are required by ``web_poet.example``.

> .. _andi: https://github.com/scrapinghub/andi

Reviewer: andi is now in install_requires, right?

.. _requests: https://requests.readthedocs.io/en/latest/user/install/

Then copy the following code into ``tutorial-project/run.py``:

.. literalinclude:: /tutorial-project/run.py
   :language: python
   :lines: 1-4, 6-12, 18

Execute that code:

..  code-block:: bash

    python tutorial-project/run.py

And the ``print(item)`` statement should output the following:

.. code-block:: python

   Book(title='The Exiled')

In this tutorial you use ``web_poet.example.get_item``, which is a simple,
incomplete implementation of the web-poet standard, built specifically for this
tutorial, for demonstration purposes. In real projects, use instead an actual
web-poet framework, like `scrapy-poet`_.

``web_poet.example.get_item`` serves to illustrate the power of web-poet: once
you have defined your page object class, a web-poet framework only needs 2
inputs from you: the URL from which you want to extract data, and the desired
output item class.

Notice that you must also call :func:`~web_poet.rules.consume_modules` once
before your first call to ``get_item``. ``consume_modules`` ensures that the

> specified Python modules are loaded. You pass ``consume_modules`` the modules

Reviewer 2: ```suggestion
specified Python modules are loaded. You pass ``consume_modules`` the module paths
```

> where you define your page object classes. Upon load,

Reviewer 2: ```suggestion
where you define your page object classes. After loading the passed module paths,
```

> paths of the modules where you define your page object classes. After loading

Reviewer 2: ```suggestion
paths of the modules where your page object classes are defined. After loading
```

:meth:`~web_poet.overrides.handle_urls` decorators register the page object

> classes that they decorate into the default registry, which ``get_item`` uses

Reviewer 2: ```suggestion
classes that they decorate into ``web_poet.default_registry``, which ``get_item`` uses
```

to determine which page object class to use based on its input parameters (URL
and item class).

Your web-poet framework can take care of everything else:

#.  It matches the input URL and item class to ``BookPage``, based on the URL
    pattern that you defined with the :meth:`~web_poet.overrides.handle_urls`
    decorator, and the return type that you declared in the page object class
    (``Book``).

#.  It inspects the inputs declared by ``BookPage``, and builds an instance of
    ``BookPage`` with the required inputs.

    ``BookPage`` is a :class:`~web_poet.pages.WebPage` subclass, and
    :class:`~web_poet.pages.WebPage` declares an attribute named ``response``
    of type :class:`~web_poet.page_inputs.HttpResponse`. Your web-poet
    framework sees this, and creates an

>     :class:`~web_poet.page_inputs.HttpResponse` object from the input URL as a

Reviewer: HttpResponse link doesn't work, not sure why

To continue this tutorial, you will need extended versions of ``Book`` and
``BookPage``, with additional fields. However, rather than editing the existing
``Book`` and ``BookPage`` classes, you will see how you can instead create new
classes that inherit them.

Append the following code to ``tutorial-project/tutorial/items.py``:

.. literalinclude:: /tutorial-project/tutorial/items.py
   :language: python
   :lines: 9-15

The code above defines a new item class, ``CategorizedBook``, that inherits the
``title`` attribute from ``Book`` and defines 2 more attributes: ``category``
and ``category_rank``.

Append the following code to
``tutorial-project/tutorial/pages/books_toscrape_com.py``:

.. literalinclude:: /tutorial-project/tutorial/pages/books_toscrape_com.py
   :language: python
   :lines: 16, 19-23, 25, 29-32

In the code above:

-   You define a new page object class: ``CategorizedBookPage``.

-   ``CategorizedBookPage`` subclasses ``BookPage``, inheriting its ``title``
    field, and defining a new one: ``category``.

    ``CategorizedBookPage`` does *not* define a ``category_rank`` field
    yet, you will add it later on. For now, the default value defined in

>     ``CategorizedBook`` for ``category_rank``, ``None``, will be used.

Reviewer 2: ```suggestion
    ``CategorizedBook`` for ``category_rank`` will be  ``None``.
```

> -   ``CategorizedBookPage`` indicates that it returns a ``CaterogizedBook``

Reviewer 2: ```suggestion
-   ``CategorizedBookPage`` indicates that it returns a ``CategorizedBook``
```

Author: My first attempt at committing this change locally:

```$ git comimt -am "CaterogizedBook → CategorizedBook"```

:facepalm:

    object.

    :class:`~web_poet.pages.WebPage` is a :ref:`generic class <generics>`,
    which is why we could use ``WebPage[Book]`` in the definition of
    ``BookPage`` to indicate ``Book`` as the output type of ``BookPage``.
    However, ``BookPage`` is not a generic class, so something like
    ``BookPage[CategorizedBook]`` would not work.

    So instead you use :class:`~web_poet.pages.Returns`, a special, generic
    class that you can inherit to re-define the output type of your page object
    subclasses.

After you update your ``tutorial-project/run.py`` script to request a
``CategorizedBook`` item:

> .. literalinclude:: /tutorial-project/run.py

Reviewer 2: Ahhh that's super cool! TIL 💯

in the list of books of the category of that book. For example,
`The Exiled`_ is 24th in the Mystery_ category, so the value of
``category_rank`` should be ``24`` for that book.

.. _The Exiled: http://books.toscrape.com/catalogue/the-exiled_247/index.html
.. _Mystery: https://books.toscrape.com/catalogue/category/books/mystery_3/

However, there is no indication of this value in the book details page. To get
this value, you need to visit the URL of the category of the book whose
data you are extracting, find the entry of that book within the grid of books
of the category, and record in which position you found it. And categories with
more than 20 books are split into multiple pages, so you may need more than 1
additional request for some books.

Extend ``CategorizedBookPage`` in
``tutorial-project/tutorial/pages/books_toscrape_com.py`` as follows:

.. literalinclude:: /tutorial-project/tutorial/pages/books_toscrape_com.py
   :language: python
   :lines: 14-15, 17, 19-26, 28-35, 39-52
   :emphasize-lines: 1-3, 9, 11-12, 18-33

In the code above:

-   You declare a new input in ``CategorizedBookPage``, ``http``, of type

>     :class:`~web_poet.page_inputs.client.HttpClient`.

Reviewer: I wonder if this point could be more clear if we don't use WebPage, and use ItemPage directly. No strong opinon though.

Author: While I think you are right about its effect on this point, I think it is good for the tutorial to cover WebPage earlier, and showcase e.g. `self.css`. No strong opinion either, though.
